# **AP155 Lab Assignment**
## Module 2: Solution of Linear and Nonlinear Equations

_Instructions_: Answer each problem as completely as you can. Discuss **all** your answers as clearly and concisely as possible.

_Scoring Criteria_: 50% - *correctness of code*; 50% - *discussion of the code and results*. Maximum score is **100 points**.



### Student Information

_Full Name (Last Name, First Name)_: Bendejo, Kevin D.\
_Student No._:2022-08528\
_Section_: TTH-TX-2

### Submission Information

_Date and Time Submitted (most recent upload)_:

**HONOR PLEDGE** I affirm that I have upheld the highest principles of honesty and integrity in my academic work and that this lab assignment is my own work.

**Sign here with your full name: Kevin Bendejo**

### Grading Information (c/o Lab Instructor)

TOTAL SCORE: **XX**/100

Score breakdown:
* Problem 1 - XX/30
* Problem 2 - XX/30
* Problem 3 - XX/40



### PROBLEM 1

**Solve for eigenvalues and eigenvectors of a symmetric matrix**

1. Use numpy's $\tt linalg$ package to find the eigenvalues and eigenvectors of the ff. symmetrix matrix *(15 pts.)*:

$ \begin{equation}
{\bf A} =
\begin{pmatrix}
8 & -2 & -20 \\
-2 & -3 & -9 \\
-20 & -9 & -3
\end{pmatrix} \end{equation} $

2. Check your answer numerically; that is, check that $\bf AV = VD$, where the $N$ x $N$ matrix $\bf V$ has the eigenvectors as its columns and $\bf D$ is the diagonal matrix with the eigenvalues $\lambda_i$ as its diagonal entries. *(15 pts.)*

In [1]:
from numpy import array
from numpy.linalg import eigh

A = array([[8.0,-2.0,-20.0],[-2.0,-3.0,-9.0],[-20.0,-9.0,-3.0]],float)
x,V = eigh(A)
print (x)
print (V)

[-21.99606195   0.16021522  23.83584674]
[[ 0.52496527  0.36793396 -0.76748685]
 [ 0.40892672 -0.89987151 -0.15169114]
 [ 0.74645187  0.2342133   0.62285932]]


In [ ]:
import numpy as np
from numpy import array
from numpy.linalg import eigh
A = array([[8,-2,-20],[-2,-3,-9],[-20,-9,-3]],float)
x,V= eigh(A)

D = array([[x[0],0,0],[0,x[1],0],[0,0,x[2]]],float)
B=np.matmul(A,V)
print (B)
print (np.matmul(V,D))


[[-11.54716868   0.05894862 -18.29369893]
 [ -8.99477756  -0.14417311  -3.61568676]
 [-16.4190016    0.03752454  14.8463793 ]]
[[-11.54716868   0.05894862 -18.29369893]
 [ -8.99477756  -0.14417311  -3.61568676]
 [-16.4190016    0.03752454  14.8463793 ]]


In [15]:
import numpy as np
from numpy import array
from numpy.linalg import eigh
A = array([[8,-2,-20],[-2,-3,-9],[-20,-9,-3]],float)
x,V= eigh(A)
for i in range (3):
  print (D[i,i])




-21.996061954235405
0.1602152189280612
23.83584673530734


### PROBLEM 2

**Solve a nonlinear equation using the relaxation method**

Consider the equation $x=1- e^{-cx}$, where $c$ is a known parameter and $x$ is unknown. This equation arises in a variety of situations, including the physics of contact processes, mathematical models of epidemics, and the theory of random graphs.

1. Write a program to solve this equation for $x$ using the relaxation method for the case $c = 2$. Calculate your solution to an accuracy of at least $10^{-6}$. *(10 pts.)*

2. Modify your program to calculate the solution for values of $c$ from 0 to 3 in steps of 0.01. *(10 pts.)*

3. Make a plot of $x$ as a function of $c$. Describe the different regimes that you see. *(10 pts.)*

In [ ]:
#in terms of similarity of digits
from math import exp
x = [0.9,1]

while round((x[-1]),7) != round((x[-2]),7):
  v = 1 - exp(-2*x[-1])
  x.append(v)

print (x[-1],x[-2], abs((x[-1])-(x[-2]))/(x[-1]))
print (len(x))


0.7968121636479626 0.7968122127708882 6.164931693713349e-08
19


In [ ]:
print (10**(-6))

1e-06


In [ ]:
from math import exp
x = [1,1]

while abs((x[-1])-(x[-2]))/(x[-1])<10**(-6):
  v = 1 - exp(-2*x[-1])
  x.append(v)

print (x[-1],x[-2])
print (len(x))


0.8646647167633873 1
3


In [ ]:
from math import exp
import numpy as np
c = np.linspace(0,3,301)
for k in range (0,301):
  x = [0.9,1]
  while round((x[-1]),7) != round((x[-2]),7):
    v = 1 - exp(-c[k]*x[-1])
    x.append(v)
  print (x[-1],x[-2])
  print (len(x))




0.0 0.0
4
9.949663315467205e-11 9.94966620204707e-09
7
6.335143520885822e-11 3.167572204532121e-09
8
7.178454497491771e-10 2.3928180437060576e-08
8
1.604752997153014e-10 4.0118833810609544e-09
9
7.610631014287605e-10 1.5221261806530606e-08
9
2.711990743087256e-09 4.519984508899455e-08
9
5.553518755974096e-10 7.933598999976255e-09
10
1.606990207569936e-09 2.0087377317068444e-08
10
4.0991773309073665e-09 4.55464150839191e-08
10
9.466140094005482e-10 9.466140316050087e-09
11
2.2185817627473625e-09 2.0168925085606304e-08
11
4.824939003889028e-09 4.0207825069416003e-08
11
1.281042161238588e-09 9.854170479606239e-09
12
2.67072020054826e-09 1.9076573098963934e-08
12
5.289644278150263e-09 3.526429526168329e-08
12
1.6030461402749552e-09 1.0019038598763075e-08
13
3.1018583257491628e-09 1.824622564150502e-08
13
5.776741862639767e-09 3.209301036033452e-08
13
1.975480778959593e-09 1.0397267380390929e-08
14
3.629735401133871e-09 1.814867722771396e-08
14
6.471005065122881e-09 3.081430988149947e-08
14

In [ ]:
import numpy as np
x = np.linspace(0,3,301)
print (x)

[0.   0.01 0.02 0.03 0.04 0.05 0.06 0.07 0.08 0.09 0.1  0.11 0.12 0.13
 0.14 0.15 0.16 0.17 0.18 0.19 0.2  0.21 0.22 0.23 0.24 0.25 0.26 0.27
 0.28 0.29 0.3  0.31 0.32 0.33 0.34 0.35 0.36 0.37 0.38 0.39 0.4  0.41
 0.42 0.43 0.44 0.45 0.46 0.47 0.48 0.49 0.5  0.51 0.52 0.53 0.54 0.55
 0.56 0.57 0.58 0.59 0.6  0.61 0.62 0.63 0.64 0.65 0.66 0.67 0.68 0.69
 0.7  0.71 0.72 0.73 0.74 0.75 0.76 0.77 0.78 0.79 0.8  0.81 0.82 0.83
 0.84 0.85 0.86 0.87 0.88 0.89 0.9  0.91 0.92 0.93 0.94 0.95 0.96 0.97
 0.98 0.99 1.   1.01 1.02 1.03 1.04 1.05 1.06 1.07 1.08 1.09 1.1  1.11
 1.12 1.13 1.14 1.15 1.16 1.17 1.18 1.19 1.2  1.21 1.22 1.23 1.24 1.25
 1.26 1.27 1.28 1.29 1.3  1.31 1.32 1.33 1.34 1.35 1.36 1.37 1.38 1.39
 1.4  1.41 1.42 1.43 1.44 1.45 1.46 1.47 1.48 1.49 1.5  1.51 1.52 1.53
 1.54 1.55 1.56 1.57 1.58 1.59 1.6  1.61 1.62 1.63 1.64 1.65 1.66 1.67
 1.68 1.69 1.7  1.71 1.72 1.73 1.74 1.75 1.76 1.77 1.78 1.79 1.8  1.81
 1.82 1.83 1.84 1.85 1.86 1.87 1.88 1.89 1.9  1.91 1.92 1.93 1.94 1.95
 1.96 

### PROBLEM 3

**Physics Application: Calculating the Lagrange point**

_Refer to Exercise 6.16 of the Newman text._ The $L_1$ Lagrange point is the point between the Earth and the Moon at which a satellite will orbit the Earth in perfect synchrony with the Moon, staying always in between the two.

Assuming circular orbits, and assuming that the Earth is much more massive than either the Moon or the satellite, the distance $r$ from the center of the Earth to the $L_1$ point satisfies
$$ \begin{equation}
\frac{GM}{r^2} - \frac{Gm}{(R-r)^2} = \omega^2 r,
\end{equation} $$
where $R$ is the distance from the Earth to the Moon, $M$ and $m$ are the Earth and Moon masses, $G$ is Newton's gravitational constant, and $\omega$ is the angular velocity of both the Moon and the satellite.

Use Newton's method to solve for the distance r from the Earth to the $L_1$ point. Compute a solution accurate to at least four significant figures.